/**:
  ros__parameters:
    ############ system configuration ############
    # Data logging
    log_to_file: true
    log_debug: false
    log_enabled:
      [
        "navigator",
        "map_projector",
        "tactic",
        "tactic.module",
        "tactic.module.live_mem_manager",
        "tactic.module.graph_mem_manager",
        "mission_planning_server",
        "state_machine",
        "pose_graph",
        "lidar.pipeline",
        "lidar.velodyne_converter",
        "lidar.honeycomb_converter",
        "lidar.preprocessing",
        "lidar.odometry_icp",
        "lidar.map_maintenance",
        "lidar.keyframe_test",
        "lidar.windowed_map_recall",
        "lidar.localization_icp",
        "path_tracker",
      ]

    # control frame from the urdf
    robot_frame: base_link
    # camera related
    camera_frame: front_xb3
    camera_topic: camera_not_used
    camera_calibration_topic: /xb3_calibration
    # lidar related
    lidar_frame: honeycomb
    lidar_topic: /points

    ############ map projection configuration ############
    map_projection:
      origin_lat: 43.78220
      origin_lng: -79.4661
      origin_theta: 1.3 # positive == clockwise
      scale: 1.0

    ############ tactic configuration ############
    tactic:
      extrapolate_odometry: true # path tracker
      localization_only_keyframe: false
      localization_skippable: true
      default_loc_cov: [1.0, 0.5, 0.5, 0.25, 0.25, 0.5]
      merge_threshold: [5.0, 3.0, 0.2]
      chain:
        min_cusp_distance: 1.5
        angle_weight: 7.0
        search_depth: 5
        search_back_depth: 10
        distance_warning: 3.0
      live_mem:
        enable: true
        lookahead_distance: 100
        window_size: 10
      map_mem:
        enable: true
        lookahead_distance: 15
        vertex_life_span: 10

      visualize: true
      vis_loc_path_offset: [0., 0., 0.]

    ############ pipeline configuration ############
    pipeline:
      type: lidar

      preprocessing: ["conversion", "filtering"]
      odometry: ["recall", "icp", "mapping", "keyframe_test"]
      localization: ["recall", "icp"]

    preprocessing:
      conversion:
        type: lidar.honeycomb_converter

        visualize: true

      filtering:
        type: lidar.preprocessing

        num_threads: 8
        frame_voxel_size: 0.1 # grid subsampling voxel size

        vertical_angle_res: 0.0132645 # vertical angle resolution in radius, equal to 0.76 degree documented in the manual
        polar_r_scale: 2.0 # polar_r_scale x vertical_angle_res = nearest neighbor search radius for normal computation
        r_scale: 4.0 # scale down point range by this value after taking log, whatever works
        h_scale: 1.54 # scale down yaw(phi) by this value so that vertical resolution ~= horizontal resolution, horizontal resolution when 5Hz spin frequence is ~1.17 degree, so 1.17 / 0.76 = 1.54

        num_sample1: 10000 # max number of sample after filtering based on planarity
        min_norm_score1: 0.95 # min planarity score

        num_sample2: 10000 # max number of sample after filtering based on planarity
        min_norm_score2: 0.2 # 0.2 is when the incident angle 5/12 * pi
        min_normal_estimate_dist: 2.0 # minimum distance to estimate normal in meters
        max_normal_estimate_angle: 0.42 # must <1/2, this value will be timed by M_PI

        cluster_num_sample: 10000 # maxnumber of sample after removing isolated points

        visualize: true

    odometry:
      recall:
        type: lidar.map_recall

        map_voxel_size: 0.3
        visualize: true

      icp:
        type: lidar.odometry_icp

        # Prior terms
        trajectory_smoothing: true
        use_constant_acc: false
        lin_acc_std_dev_x: 1.0
        lin_acc_std_dev_y: 0.1
        lin_acc_std_dev_z: 0.1
        ang_acc_std_dev_x: 0.1
        ang_acc_std_dev_y: 0.1
        ang_acc_std_dev_z: 1.0

        # ICP parameters
        num_threads: 8
        first_num_steps: 3
        initial_max_iter: 6
        initial_num_samples: 12000
        initial_max_pairing_dist: 1.5
        initial_max_planar_dist: 1.0
        refined_max_iter: 50
        refined_num_samples: 12000
        refined_max_pairing_dist: 1.0
        refined_max_planar_dist: 0.3

        min_matched_ratio: 0.5

      mapping:
        type: lidar.map_maintenance

        map_voxel_size: 0.3

        # dynamic objects removal
        horizontal_resolution: 0.02042
        vertical_resolution: 0.01326
        min_num_observations: 1
        max_num_observations: 5

        visualize: true

      keyframe_test:
        type: lidar.keyframe_test

        min_translation: 1.0
        min_rotation: 15.0
        max_translation: 3.0
        max_rotation: 30.0

        min_matched_points_ratio: 0.6

        max_num_points: 100000

    localization:
      recall:
        type: lidar.windowed_map_recall

        single_exp_map_voxel_size: 0.3
        multi_exp_map_voxel_size: 0.3

        remove_short_term_dynamic: true
        short_term_min_num_observations: 1 # must be positive otherwise undefined behavior
        short_term_min_movability: 0.5

        depth: 4
        num_additional_exps: 5
        long_term_min_num_observations: 4

        visualize: true

      icp:
        type: lidar.localization_icp

        use_pose_prior: true

        # ICP parameters
        num_threads: 8
        first_num_steps: 3
        initial_max_iter: 10
        initial_num_samples: 12000
        initial_max_pairing_dist: 1.5
        initial_max_planar_dist: 1.0
        refined_max_iter: 50
        refined_num_samples: 12000
        refined_max_pairing_dist: 1.0
        refined_max_planar_dist: 0.3

        min_matched_ratio: 0.5
